src/javascript/class-scanner.ts

```diff
@@ -594,70 +668,72 @@ export function extractPropertiesFromConstructor(
     return properties;
   }
   for (const method of astNode.body.body) {
-    if (!babel.isClassMethod(method) || method.kind !== 'constructor') {
+    if (!babel.isClassMethod(method)) {
       continue;
     }
-    const constructor = method;
-    for (const statement of constructor.body.body) {
-      if (!babel.isExpressionStatement(statement)) {
-        continue;
-      }
-      let name;
-      let astNode;
-      let defaultValue;
-      if (babel.isAssignmentExpression(statement.expression)) {
-        // statements like:
-        // /** @public The foo. */
-        // this.foo = baz;
-        name = getPropertyNameOnThisExpression(statement.expression.left);
-        astNode = statement.expression.left;
-        defaultValue = generate(statement.expression.right).code;
-      } else if (babel.isMemberExpression(statement.expression)) {
-        // statements like:
-        // /** @public The foo. */
-        // this.foo;
-        name = getPropertyNameOnThisExpression(statement.expression);
-        astNode = statement;
-      } else {
-        continue;
+
+    if (method.kind === 'constructor') {
+      const constructor = method;
+      for (const statement of constructor.body.body) {
+        if (!babel.isExpressionStatement(statement)) {
+          continue;
+        }
+        const prop = extractPropertyFromExpressionStatement(statement, document);
+        if (!prop) {
+          continue;
+        }
+        properties.set(prop.name, prop);
       }
-      if (name === undefined) {
+    } else if (method.kind === 'get' || method.kind === 'set') {
+      const prop = extractPropertyFromMethod(method, document);
+      if (!prop) {
         continue;
       }
-      const comment = esutil.getAttachedComment(statement);
-      const jsdocAnn =
-          comment === undefined ? undefined : jsdoc.parseJsdoc(comment);
-      if (!jsdocAnn || jsdocAnn.tags.length === 0) {
-        // The comment only counts if there's a jsdoc annotation in there
-        // somewhere.
-        // Otherwise it's just an assignment, maybe to a property in a
-        // super class or something.
-        continue;
+
+      let existing = properties.get(prop.name);
+
+      if (!existing) {
+        prop.readOnly = (method.kind === 'get');
```

Comment:
Seems like we should have an explicit way to indicate that something is a getter or setter.

Reply:
IMO from the outside IMO the only thing that matters is whether a property is readonly or not. It's an implementation detail whether a field is a property or a getter/setter pair.

Reply:
At least I hope the differences that JavaScript has aren't ever important

Reply:
Fair enough. I was thinking we'd want to know for the generated TypeScript declarations, but in fact we don't (actually it's not even valid to declare a getter in an ambient context, only in an implementation).

Reply:
this is majorly wrong at the min.

reason being, jsdoc states you should only annotate one of the two (the get or the set). so there should never be an `existing`. what we actually need to do is assume the one with an annotation is the correct one, then look for the other at that point.